import/index_java/src/org/vufind/index/FormatCalculator.java

```diff
@@ -199,17 +207,28 @@ protected String getFormatFrom007(char formatCode, String formatString) {
      *
      * @param Record record
      * @param char bibLevel
-     * @param char formatCode
      * @param ControlField marc008
      * @param boolean couldBeBook
+     * @param List formatCodes007
      * @return String
      */
-    protected String getFormatFromBibLevel(Record record, char bibLevel, char formatCode, ControlField marc008, boolean couldBeBook) {
+    protected String getFormatFromBibLevel(Record record, char bibLevel, ControlField marc008, boolean couldBeBook, List formatCodes007) {
         switch (bibLevel) {
             // Monograph
             case 'm':
                 if (couldBeBook) {
-                    return (formatCode == 'c') ? "eBook" : "Book";
+                    // Check 008/23 Form of item
+                    switch (get008Value(marc008, 23)) {
+                        case 'o': // Online
+                        case 'q': // Direct electronic
+                        case 's': // Electronic
+                            return "eBook";
+                        default: break;
+                    }
+                    // Fall-back on 007 if 008 is missing
+                    // Note: relying on 007 here is not ideal as it is repeatable
```

Comment:
Would appreciate your advice here @demiankatz  (and same below at line 258).

I left this in mostly for backwards compatibility but it doesn't really make sense to me:

- the 007 is repeatable
- Unlike the 008 which tries to encode the main "thing" being described, the 007 can provide additional detail about the main thing being described, or about part of the main thing being described, or about "accompanying material"
- So in this case, the main thing being described is a good old fashioned printed book but there might be a 007 describing the fact that there is something electronic remote 007cr (e.g. a version of the book online, but presumably also some accompanying material that is online)
- in the latter case, we've now classed the good old fashioned book as an ebook incorrectly
- you can actually see this elsewhere on line 112 where we ignore 007cr entries because they are so value and might end up with lots of erroneous "Electronic" formats

As we're now looking in the 008 to determine if the "main thing" is electronic can we just drop this 007 check here (obviously anything returned separately by getFormatFrom007 will still be there)

Reply:
That sounds reasonable to me. We can make the change and see what it does to our tests, in any case!